File: C++/src/Vect3.cpp
/*
 * Vect3.cpp
 *
 * Contact: Cesar Munoz (cesar.a.munoz@nasa.gov)
 * NASA LaRC
 *
 * 3-D vectors.
 *
 * Copyright (c) 2011-2016 United States Government as represented by
 * the National Aeronautics and Space Administration.  No copyright
 * is claimed in the United States under Title 17, U.S.Code. All Other
 * Rights Reserved.
 */

#include "Vect3.h"
#include "Units.h"
#include <limits>
#include "string_util.h"
#include "format.h"
#include "Constants.h"
#include "Util.h" // NaN def

namespace larcfm {

Vect3::Vect3(const double xx, const double yy, const double zz) : x(xx), y(yy), z(zz) {
}

Vect3::Vect3(const Vect2&v, const double vz) : x(v.x), y(v.y), z(vz) {
}

Vect3 Vect3::makeXYZ(double x, std::string ux, double y, std::string uy, double z, std::string uz) {
	return Vect3(Units::from(ux,x),Units::from(uy,y),Units::from(uz,z));
}

Vect3 Vect3::mkXYZ(double x, double y, double z) {
	return Vect3(x,y,z);
}

Vect3 Vect3::mkX(double nx) {
	return Vect3(nx,y,z);
}
Vect3 Vect3::mkY(double ny) {
	return Vect3(x,ny,z);
}
Vect3 Vect3::mkZ(double nz) {
	return Vect3(x,y,nz);
}

bool Vect3::isZero() const {
	return x == 0.0 && y == 0.0 && z == 0.0;
}

bool Vect3::almostEquals(const Vect3& v) const {
	return Util::almost_equals(x,v.x) && Util::almost_equals(y,v.y) && Util::almost_equals(z,v.z);
}

bool Vect3::almostEquals(const Vect3& v, INT64FM maxUlps) const {
	return Util::almost_equals(x,v.x,maxUlps) && Util::almost_equals(y,v.y,maxUlps) && Util::almost_equals(z,v.z,maxUlps);
}

bool Vect3::within_epsilon(const Vect3& v2, double epsilon) const {
	if (std::abs(x - v2.x) > epsilon) return false;
	if (std::abs(y - v2.y) > epsilon) return false;
	if (std::abs(z - v2.z) > epsilon) return false;
	return true;
}


Vect3 Vect3::operator + (const Vect3& v) const {
	return this->Add(v);
}

Vect3 Vect3::operator - (const Vect3& v) const {
	return this->Sub(v);
}

Vect3 Vect3::operator - () const {
	return Vect3(-x,-y,-z);
}

Vect3 Vect3::operator * (const double k) const {
	return this->Scal(k);
}

double Vect3::operator * (const Vect3& v) const { // Dot product
	return dot(v.x,v.y,v.z);
}

bool Vect3::operator == (const Vect3& v) const {  // strict equality
	return x==v.x && y==v.y && z==v.z;
}

bool Vect3::operator != (const Vect3& v) const {  // strict disequality
	return x!=v.x || y!=v.y || z!=v.z;
}

Vect2 Vect3::vect2() const {
	return Vect2(x,y);
}

Vect3 Vect3::Hat() const {
	double n = norm();
	if ( n == 0.0) { // this is only checking the divide by zero case, so an exact comparison is correct.
		return ZERO();
	}
	return Vect3(x/n, y/n, z/n);
}

Vect3 Vect3::Hat2D() const {
	return Vect3(x,y,0.0).Hat();
}

Vect3 Vect3::cross(const Vect3& v) const {
	return Vect3(this->y*v.z - this->z*v.y, this->z*v.x - this->x*v.z, this->x*v.y - this->y*v.x);
}

bool Vect3:: parallel(const Vect3& v) const {
	return cross(v).almostEquals(Vect3::ZERO());
}

Vect3 Vect3::Add(const Vect3& v) const{
	return Vect3(x+v.x, y+v.y, z+v.z);
}

Vect3 Vect3::Sub(const Vect3& v) const {
	return Vect3(x-v.x,y-v.y,z-v.z);
}

Vect3 Vect3::Neg() const {
	return Vect3(-x,-y,-z);
}

Vect3 Vect3::Scal(double k) const {
	return Vect3(k*x,k*y,k*z);
}

Vect3 Vect3::ScalAdd(const double k, const Vect3& v) const {
	return Vect3(k*x+v.x, k*y+v.y, k*z+v.z);
}

Vect3 Vect3::AddScal(double k, const Vect3& v) const {
	return Vect3(x+k*v.x, y+k*v.y, z+k*v.z);
}

Vect3 Vect3::PerpR() const {
	return Vect3(y,-x, 0);
}

Vect3 Vect3::PerpL() const {
	return Vect3(-y,x, 0);
}

Vect3 Vect3::linear(const Vect3& v, double t) const {
	return Vect3(x+v.x*t, y+v.y*t, z+v.z*t);
}

Vect3 Vect3::linearByDist(double track, double d) const {
	double gs = 100;
	//Velocity v = Velocity::mkTrkGsVs(track,gs,0.0);
	Vect3 v = Vect3(gs*sin(track),gs*cos(track),0.0);
	double dt = d/gs;
	return linear(v,dt);
 }


double Vect3::dot(const double x, const double y, const double z) const {
	return this->x*x + this->y*y + this->z*z;
}

double Vect3::dot(const Vect3& v) const {
	return dot(v.x, v.y, v.z);
}

double Vect3::sqv() const {
	return dot(x,y,z);
}

double Vect3::norm() const {
	return sqrt_safe(sqv());
}

double Vect3::cyl_norm(const double d, const double h) const {
	return std::max(vect2().sqv()/sq(d),sq(z/h));
}

double Vect3::distanceH(const Vect3& w) const {
	Vect2 v = Vect2(x,y);
	return (v-w.vect2()).norm();
}
double Vect3::distanceV(const Vect3& w) const {
	return z - w.z;
}

std::string Vect3::toString() const {
	return toString(Constants::get_output_precision());
}

std::string Vect3::toString(int precision) const {
	return formatXYZ(precision,"(",", ",")");
}

std::string Vect3::toStringNP(const std::string& xunit, const std::string& yunit, const std::string& zunit, int prec) const {
	return FmPrecision(Units::to(xunit,x),prec)+", "+FmPrecision(Units::to(yunit,y),prec)+", " +
			FmPrecision(Units::to(zunit,z),prec);
}

std::string Vect3::formatXYZ(int prec, const std::string& pre, const std::string& mid, const std::string& post) const {
	return pre+FmPrecision(x,prec)+mid+FmPrecision(y,prec)+mid+FmPrecision(z,prec)+post;
}

std::string Vect3::toPVS(int prec) const {
	return "(# x:= "+FmPrecision(x,prec)+", y:= "+FmPrecision(y,prec)+", z:= "+FmPrecision(z,prec)+" #)";
}

/**
 * Returns true if the current vector has an "invalid" value
 */
bool Vect3::isInvalid() const {
	return x != x || y != y || z != z;
}

const Vect3& Vect3::ZERO() {
	static Vect3* v = new Vect3(0,0,0);
	return *v;
}

const Vect3& Vect3::INVALID() {
	static Vect3* v = new Vect3(NaN, NaN, NaN);
	return *v;
}

Vect3 Vect3::parse(const std::string& s) {
	std::vector<std::string> fields = split(s, Constants::wsPatternParens);
	while (fields.size() > 0 && equals(fields[0], "")) {
		fields.erase(fields.begin());
	}
	if (fields.size() == 3) {
		return Vect3(Util::parse_double(fields[0]),Util::parse_double(fields[1]),Util::parse_double(fields[2]));
	} else if (fields.size() == 6) {
		return Vect3(Units::from(Units::clean(fields[1]),Util::parse_double(fields[0])),
				Units::from(Units::clean(fields[3]),Util::parse_double(fields[2])),
				Units::from(Units::clean(fields[5]),Util::parse_double(fields[4])));
	}
	return Vect3::INVALID();
}

}

File: Java/src/gov/nasa/larcfm/Util/Vect3.java
/*
 * Vect3.java 
 * 
 * 3-D vectors.
 * 
 * Copyright (c) 2011-2016 United States Government as represented by
 * the National Aeronautics and Space Administration.  No copyright
 * is claimed in the United States under Title 17, U.S.Code. All Other
 * Rights Reserved.
 */

package gov.nasa.larcfm.Util;

import java.util.Arrays;

/**
 * 3-Dimensional mathematical vectors.
 */
public class Vect3 {
  
  /** A zero vector */
  public static final Vect3 ZERO = new Vect3();

  /** An invalid Vect3.  Note that this is not necessarily equal to other invalid Vect3s -- use the isInvalid() test instead. */
  public static final Vect3 INVALID = new Vect3(Double.NaN,Double.NaN,Double.NaN);

  /** The x-component */
  public final double x;

  /** The y-component */
  public final double y;

  /** The z-component */
  public final double z;

  /**
   * Creates a zero vector.
   */
  private Vect3() {
    this(0.0,0.0,0.0);
  }

  /**
   * Creates a vector that is an extension of v
   */
  public Vect3(Vect2 v, double z) {
    this(v.x,v.y,z);
  }

  /**
   * Creates a new vector with coordinates (<code>x</code>,<code>y</code>,<code>z</code>).
   * 
   * @param x Real value [internal units]
   * @param y Real value [internal units]
   * @param z Real value [internal units]
   */
  public Vect3(double x, double y, double z) {
    this.x = x;
    this.y = y;
    this.z = z;
  }

  /**
   * Creates a new vector with coordinates (<code>x</code>,<code>y</code>,<code>z</code>) in specified units.
   * 
   * @param x  Real value [ux]
   * @param ux Units x
   * @param y  Real value [uy]
   * @param ux Units y
   * @param z  Real value [uz]
   * @param uz Units z
   */
  public static Vect3 makeXYZ(double x, String ux, double y, String uy, double z, String uz) {
    return new Vect3(Units.from(ux,x),Units.from(uy,y),Units.from(uz,z));
  }

  /**
   * Creates a new vector with coordinates (<code>x</code>,<code>y</code>,<code>z</code>) in internal units.
   * 
   * @param x Real value [internal units]
   * @param y Real value [internal units]
   * @param z Real value [internal units]
   */
  public static Vect3 mkXYZ(double x, double y, double z) {
    return new Vect3(x,y,z);
  }

  public Vect3 mkX(double nx) {
    return mkXYZ(nx, y ,z);
  }

  public Vect3 mkY(double ny) {
    return mkXYZ(x, ny ,z);
  }

  public Vect3 mkZ(double nz) {
    return mkXYZ(x, y , nz);
  }

  /** The x coordinate */
  public double x() {
    return x;
  }

  /** The y coordinate */
  public double y() {
    return y;
  }

  /** The z coordinate */
  public double z() {
    return z;
  }

  /**
   * Zero constant.
   */
  public static Vect3 Zero() {
    return new Vect3();
  }

  @Override
  public int hashCode() {
    final int prime = 31;
    int result = 1;
    long temp;
    temp = Double.doubleToLongBits(x);
    result = prime * result + (int) (temp ^ (temp >>> 32));
    temp = Double.doubleToLongBits(y);
    result = prime * result + (int) (temp ^ (temp >>> 32));
    temp = Double.doubleToLongBits(z);
    result = prime * result + (int) (temp ^ (temp >>> 32));
    return result;
  }

  @Override
  public boolean equals(Object obj) {
    if (this == obj)
      return true;
    if (obj == null)
      return false;
    //if (getClass() != obj.getClass())
    if (!(obj instanceof Vect3))
      return false;
    Vect3 other = (Vect3) obj;
    if (Double.doubleToLongBits(x) != Double.doubleToLongBits(other.x))
      return false;
    if (Double.doubleToLongBits(y) != Double.doubleToLongBits(other.y))
      return false;
    if (Double.doubleToLongBits(z) != Double.doubleToLongBits(other.z))
      return false;
    return true;
  }

  /**
   * Checks if vector is zero.
   * 
   * @return <code>true</code>, if <code>this</code> vector is zero.
   */
  public boolean isZero() {
    return x == 0.0 && y == 0.0 && z == 0.0;
  }

  /**
   * Checks if vectors are almost equal.
   * 
   * @param v Vector
   * 
   * @return <code>true</code>, if <code>this</code> vector is almost equal 
   * to <code>v</code>.
   */
  public boolean almostEquals(Vect3 v) {
    return Util.almost_equals(x,v.x) && Util.almost_equals(y,v.y) && 
        Util.almost_equals(z,v.z);
  }

  /**
   * Checks if vectors are almost equal.
   * 
   * @param v Vector
   * @param maxUlps unit of least precision
   * 
   * @return <code>true</code>, if <code>this</code> vector is almost equal 
   * to <code>v</code>.
   */
  public boolean almostEquals(Vect3 v, long maxUlps) {
    return Util.almost_equals(x, v.x, maxUlps) && Util.almost_equals(y, v.y, maxUlps) && 
        Util.almost_equals(z, v.z, maxUlps);
  }

  public boolean within_epsilon(Vect3 v2, double epsilon) {
    //f.pln(" $$$ Vect3.within_epsilon: v1 = "+(new Vect3(x,y,z))+ " v2 = "+v2+" epsilon = "+epsilon );
    //f.pln(" $$$ Vect3.within_epsilon: z = "+z+" v2.z = "+v2.z);
    if (Math.abs(x - v2.x) > epsilon) return false;
    if (Math.abs(y - v2.y) > epsilon) return false;
    if (Math.abs(z - v2.z) > epsilon) return false;
    //f.pln(" $$$ Vect3.within_epsilon: return true");
    return true;  
  }

  /**
   * Dot product.
   * 
   * @param v Vector
   * 
   * @return the dot product of <code>this</code> vector and <code>v</code>.
   */
  public double dot(Vect3 v) {
    return dot(v.x,v.y,v.z);
  }

  /**
   * Dot product.
   * 
   * @param x Real value
   * @param y Real value
   * @param z Real value
   * 
   * @return the dot product of <code>this</code> vector and (<code>x</code>,<code>y</code>,<code>z</code>).
   */
  public double dot(double x, double y, double z) {
    return this.x*x + this.y*y + this.z*z;
  }

  /**
   * Square.
   * 
   * @return the dot product of <code>this</code> vector with itself.
   */
  public double sqv() {
    return dot(x,y,z); // dot product of this vector with itself
  }

  /**
   * Norm.
   * 
   * @return the norm of of <code>this</code> vector.
   */
  public double norm() {
    return Util.sqrt_safe(sqv());
  }

  /**
   * Make a unit vector from the current vector.  If it is a zero vector, then a copy is returned.
   * @return the unit vector
   */
  public Vect3 Hat() {
    double n = norm();
    if (n == 0.0) { // this is only checking the divide by zero case, so an exact comparison is correct.
      return this;
    }
    return new Vect3(x/n, y/n, z/n);
  }

  public Vect3 Hat2D() {
	  return mkZ(0.0).Hat();
  }
  
  /**
   * Cross product.
   * 
   * @param v Vector
   * 
   * @return the cross product of <code>this</code> vector and <code>v</code>.
   */
  public Vect3 cross(Vect3 v) {
    return new Vect3(this.y*v.z - this.z*v.y, this.z*v.x - this.x*v.z, this.x*v.y - this.y*v.x);
  }

  public boolean parallel(Vect3 v) {
    return cross(v).almostEquals(Vect3.ZERO); 
  }

  /**
   * 2-Dimensional projection.
   *
   * @return the 2-dimensional projection of <code>this</code>.
   */
  public Vect2 vect2() {
    return new Vect2(x,y);
  }

  /**
   * Vector addition.
   * 
   * @param v Vector
   * 
   * @return the vector addition of <code>this</code> vector and <code>v</code>.
   */
  public Vect3 Add(Vect3 v) {
    return new Vect3(x+v.x, y+v.y, z+v.z);
  }

  /**
   * Vector subtraction.
   * 
   * @param v Vector
   * 
   * @return the vector subtraction of <code>this</code> vector and <code>v</code>.
   */
  public Vect3 Sub(Vect3 v) {
    return new Vect3(x-v.x, y-v.y, z-v.z);
  }

  /**
   * Vector negation.
   * 
   * @return the vector negation of of <code>this</code> vector.
   */
  public Vect3 Neg() {
    return new Vect3(-x,-y,-z);
  }

  /**
   * Scalar multiplication.
   * 
   * @param k Real value
   * 
   * @return the vector scalar multiplication of <code>this</code> vector and <code>k</code>.
   */
  public Vect3 Scal(double k) {
    return new Vect3(k*x, k*y, k*z);
  }

  /**
   * Scalar and addition multiplication. Compute: k*<code>this</code> + v
   * 
   * @param k Real value
   * @param v Vector
   * 
   * @return the scalar multiplication <code>this</code> vector and <code>k</code>, followed by an
   * addition to vector <code>v</code>.
   */
  public Vect3 ScalAdd(double k, Vect3 v) {
    return new Vect3(k*x+v.x, k*y+v.y, k*z+v.z);
  }

  /**
   * Addition and scalar multiplication.  Compute: this + k*<code>v</code>;
   * 
   * @param k real value
   * @param v vector
   * 
   * @return the addition of <code>this</code> vector to <code>v</code> scaled by <code>k</code>.
   */
  public Vect3 AddScal(double k, Vect3 v) {
    return new Vect3(x+k*v.x, y+k*v.y, z+k*v.z);
  }

  /**
   * Right perpendicular, z-component set to 0
   * 
   * @return the right perpendicular of <code>this</code> vector, i.e., (<code>y</code>, <code>-x</code>).
   */
  public Vect3 PerpR() {
    return new Vect3(y,-x,0);
  }

  /**
   * Left perpendicular, z-component set to 0
   * 
   * @return the left perpendicular of <code>this</code> vector, i.e., (<code>-y</code>, <code>x</code>).
   */
  public Vect3 PerpL() {
    return new Vect3(-y,x,0);
  }

  /**
   * Calculates position after t time units in direction and magnitude of velocity v
   * @param v    velocity
   * @param t    time
   * @return the new position
   */
  public Vect3 linear(Vect3 v, double t) {
    return new Vect3(x+v.x*t, y+v.y*t, z+v.z*t);
  }
  
  /**
   * Calculates position after t time units in direction and magnitude of velocity v
   * @param track  
   * @param t       time
   * @return the new position (horizontal only)
   */
  public Vect3 linearByDist(double track, double d) {
	double anySpeed = 100;
	Velocity v = Velocity.mkTrkGsVs(track,anySpeed,0.0);
	double dt = d/anySpeed;
	return linear(v,dt);
  }


  /** 3-D time of closest point of approach 
   * if time is negative or velocities are parallel returns 0
   */
  public static double tcpa (Vect3 so, Vect3 vo, Vect3 si, Vect3 vi) {
    double t;
    Vect3 s = so.Sub(si);
    Vect3 v = vo.Sub(vi);
    double nv = v.sqv();
    if (nv > 0) 
      t = Util.max(0,-s.dot(v)/nv);
    else 
      t = 0;
    return t;
  }// tcpa

  /** 3-D distance at time of closest point of approach
   **/
  public static double dcpa(Vect3 so, Vect3 vo, Vect3 si, Vect3 vi) {
    double t = tcpa(so,vo,si,vi);
    Vect3 s = so.Sub(si);
    Vect3 v = vo.Sub(vi);
    Vect3 st = s.AddScal(t,v);
    return st.norm();
  }// dcpa

  /**
   * Returns true if the current vector has an "invalid" value
   */
  public boolean isInvalid() {
    return Double.isNaN(x) || Double.isNaN(y) || Double.isNaN(z);
  }

  /**
   * Cylindrical norm.
   * @param d Radius of cylinder
   * @param h Half-height of cylinder
   * @return the cylindrical distance of <code>this</code>. The cylindrical distance is
   * 1 when <code>this</code> is at the boundaries of the cylinder. 
   */
  public double cyl_norm(double d, double h) {
    return Util.max(vect2().sqv()/Util.sq(d),Util.sq(z/h));
  }

  /**
   * Compare two vectors: return true iff delta is within specified limits 
   * 
   * @param v a Vect3
   * @param maxX
   * @param maxY
   * @param maxZ
   * @return true iff each component of the vector is within the given bound.
   */
  public boolean compare(Vect3 v, double maxX, double maxY, double maxZ) {
    if (Math.abs(v.x - x) > maxX) return false;
    if (Math.abs(v.y - y) > maxY) return false;
    if (Math.abs(v.z - z) > maxZ) return false;
    return true;
  }

  /** The horizontal distance between this vector and the given vector, essentially same as v.Sub(w).vect2().norm() */
  public double distanceH(Vect3 w) {
    Vect2 v = new Vect2(x,y);
    return v.Sub(w.vect2()).norm(); 
  }

  /** The vertical distance between this vector and the given vector, essentially same as v.z - w.z */
  public double distanceV(Vect3 w) {
    return z - w.z;
  }

  /** A string representation of this vector */
  public String toString() {
    return toString(Constants.get_output_precision());
  }

	/** A string representation of this vector */
  public String toString(int precision) {
    return formatXYZ(precision,"(",", ",")");
  }

  public String toStringNP(String xunit, String yunit, String zunit, int prec) {
    return f.FmPrecision(Units.to(xunit, x), prec) + ", " + f.FmPrecision(Units.to(yunit, y), prec) + ", " 	+ f.FmPrecision(Units.to(zunit, z), prec);
  }

  public String formatXYZ(int prec, String pre, String mid, String post) {
    return pre+f.FmPrecision(x,prec)+mid+f.FmPrecision(y,prec)+mid+f.FmPrecision(z,prec)+post;
  }

  public String toPVS(int prec) {
    return "(# x:= "+f.FmPrecision(x,prec)+", y:= "+f.FmPrecision(y,prec)+", z:= "+f.FmPrecision(z,prec)+" #)";
  }

  /** 
   * This parses a space or comma-separated string as a Vect3 (an inverse to the toString method).  If three 
   * bare values are present, then it is interpreted as internal units.
   * If there are 3 value/unit pairs then each values is interpreted wrt the appropriate unit.  If the string 
   * cannot be parsed, an INVALID value is returned. 
   * */
  public static Vect3 parse(String str) {
    String[] fields = str.split(Constants.wsPatternParens);
    if (fields[0].equals("")) {
      fields = Arrays.copyOfRange(fields,1,fields.length);
    }
    try {
      if (fields.length == 3) {
        return new Vect3(Double.parseDouble(fields[0]),Double.parseDouble(fields[1]),Double.parseDouble(fields[2]));
      } else if (fields.length == 6) {
        return new Vect3(Units.from(Units.clean(fields[1]),Double.parseDouble(fields[0])),
            Units.from(Units.clean(fields[3]),Double.parseDouble(fields[2])),
            Units.from(Units.clean(fields[5]),Double.parseDouble(fields[4])));
      }
    } catch (Exception e) {}
    return Vect3.INVALID;
  }

}
